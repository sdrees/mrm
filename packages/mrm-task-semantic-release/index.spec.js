jest.mock('fs');
jest.mock('mrm-core/src/util/log', () => ({
	added: jest.fn(),
}));
jest.mock('mrm-core/src/npm', () => ({
	uninstall: jest.fn(),
}));

const { uninstall } = require('mrm-core');
const { getConfigGetter } = require('mrm');
const vol = require('memfs').vol;
const task = require('./index');

const console$log = console.log;

const stringify = o => JSON.stringify(o, null, '  ');

const travisYml = `language: node_js
node_js:
  - 8
`;
const packageJson = stringify({
	name: 'unicorn',
	scripts: {
		'semantic-release': 'semantic-release',
	},
});
const readmeMd = '# Unicorn';

beforeEach(() => {
	console.log = jest.fn();
});

afterEach(() => {
	vol.reset();
	uninstall.mockClear();
	console.log = console$log;
});

it('should add semantic-release', () => {
	vol.fromJSON({
		'/.travis.yml': travisYml,
		'/package.json': packageJson,
		'/Readme.md': readmeMd,
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()).toMatchSnapshot();
});

it('should add custom config to package.json', () => {
	vol.fromJSON({
		'/.travis.yml': travisYml,
		'/package.json': packageJson,
	});

	task(
		getConfigGetter({
			semanticConfig: { pizza: 42 },
		})
	);

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
});

it('should remove custom config from package.json', () => {
	vol.fromJSON({
		'/.travis.yml': travisYml,
		'/package.json': stringify({
			name: 'unicorn',
			scripts: {
				'semantic-release': 'semantic-release',
			},
			release: {
				analyzeCommits: 'semantic-release-tamia/analyzeCommits',
			},
		}),
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
});

it('should add custom arguments to semantic-release command', () => {
	vol.fromJSON({
		'/.travis.yml': travisYml,
		'/package.json': packageJson,
	});

	task(
		getConfigGetter({
			semanticArgs: '--arg val',
		})
	);

	expect(vol.toJSON()['/.travis.yml']).toMatchSnapshot();
});

it('should install extra dependencies on CI', () => {
	vol.fromJSON({
		'/.travis.yml': travisYml,
		'/package.json': packageJson,
	});

	task(
		getConfigGetter({
			semanticPeerDependencies: ['pizza'],
		})
	);

	expect(vol.toJSON()['/.travis.yml']).toMatchSnapshot();
});

it('should crate semantic-release config and install extra dependencies when a preset is defined', () => {
	vol.fromJSON({
		'/.travis.yml': travisYml,
		'/package.json': packageJson,
	});

	task(
		getConfigGetter({
			semanticPreset: 'pizza',
		})
	);

	expect(vol.toJSON()['/.travis.yml']).toMatchSnapshot();
	expect(vol.toJSON()['/.releaserc.json']).toMatchSnapshot();
});

it('should remove the official semantic-release runner', () => {
	vol.fromJSON({
		'/.travis.yml': `language: node_js
node_js:
  - 8
after_success:
  - bash <(curl -s https://codecov.io/bash)
  - npm run semantic-release
`,
		'/package.json': packageJson,
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()['/.travis.yml']).toMatchSnapshot();
	expect(uninstall).toBeCalledWith(['semantic-release', 'travis-deploy-once']);
});

it('should throw when .travis.yml not found', () => {
	vol.fromJSON({
		'/package.json': packageJson,
	});

	const fn = () => task(getConfigGetter({}));

	expect(fn).toThrowError('Run travis task');
});
