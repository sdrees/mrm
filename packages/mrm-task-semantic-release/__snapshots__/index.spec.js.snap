// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add custom arguments to semantic-release command 1`] = `
"language: node_js
node_js:
  - 8
after_success:
  - npx semantic-release --arg val
branches:
  except:
    - /^v\\\\d+\\\\.\\\\d+\\\\.\\\\d+$/
"
`;

exports[`should add custom config to package.json 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {},
  \\"release\\": {
    \\"pizza\\": 42
  }
}"
`;

exports[`should add semantic-release 1`] = `
Object {
  "/.travis.yml": "language: node_js
node_js:
  - 8
after_success:
  - npx semantic-release
branches:
  except:
    - /^v\\\\d+\\\\.\\\\d+\\\\.\\\\d+$/
",
  "/Readme.md": "# Unicorn

[![npm](https://img.shields.io/npm/v/unicorn.svg)](https://www.npmjs.com/package/unicorn)",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {}
}",
}
`;

exports[`should crate semantic-release config and install extra dependencies when a preset is defined 1`] = `
"language: node_js
node_js:
  - 8
after_success:
  - npm install --no-save conventional-changelog-pizza
  - npx semantic-release
branches:
  except:
    - /^v\\\\d+\\\\.\\\\d+\\\\.\\\\d+$/
"
`;

exports[`should crate semantic-release config and install extra dependencies when a preset is defined 2`] = `
"{
  \\"plugins\\": [
    [
      \\"@semantic-release/commit-analyzer\\",
      {
        \\"preset\\": \\"pizza\\"
      }
    ],
    [
      \\"@semantic-release/release-notes-generator\\",
      {
        \\"preset\\": \\"pizza\\"
      }
    ],
    \\"@semantic-release/npm\\",
    \\"@semantic-release/github\\"
  ]
}"
`;

exports[`should install extra dependencies on CI 1`] = `
"language: node_js
node_js:
  - 8
after_success:
  - npm install --no-save pizza
  - npx semantic-release
branches:
  except:
    - /^v\\\\d+\\\\.\\\\d+\\\\.\\\\d+$/
"
`;

exports[`should remove custom config from package.json 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {}
}"
`;

exports[`should remove the official semantic-release runner 1`] = `
"language: node_js
node_js:
  - 8
after_success:
  - 'bash <(curl -s https://codecov.io/bash)'
  - npx semantic-release
branches:
  except:
    - /^v\\\\d+\\\\.\\\\d+\\\\.\\\\d+$/
"
`;
